R/nse-inject.R

```diff
@@ -385,6 +388,43 @@ NULL
 #' usethis::use_package("glue", "Imports")
 #' ```
 #'
+#' @section Wrapping `englue()`:
+#'
+#' You can provide englue semantics to a string by supplying `env`.
+#' In this example we create a variant of `englue()` that supports a
+#' special `.qux` pronoun by:
+#'
+#' - Creating an environment `masked_env` that inherits from the user
+#'   env, the one where their data lives.
+#'
+#' - Overriding the `error_arg` and `error_call` arguments to point to
+#'   our own argument name and call environment. This pattern is
+#'   slightly different from usual error context passing because
+#'   `englue()` is a backend function that uses its own error context
+#'   by default (and not a checking function that uses _your_ error
+#'   context by default).
+#'
+#' ```{r}
+#' my_englue <- function(text) {
```

Comment:
I think `my_englue()` should also expose `env`.

I think this is how we will end up documenting `.envir` in glue, like https://github.com/tidyverse/glue/issues/281#issuecomment-1363096413

Reply:
I guess the main point is ensuring that `caller_env()` gets used and passed through somehow, so it isn't that important... but it still might be "best practice"?

Or maybe you can say:

> If you are creating a low level package on top of `englue()`, you should also consider exposing `env`, `error_arg` and `error_call` in your `englue()` wrapper so users can wrap your wrapper.

Reply:
I like your note about exposing these arguments to allow wrappers.

There are cases where this would be unwieldy though, e.g. in the tidyr case.

By the way, maybe we should consider an `enquo()`-based interface for englue? We'd require literal strings as inputs and to wrap englue you'd use `{{`.

Reply:
This would have the advantage of making it easy to wrap `names_glue` if the tidyr function adds englue support, without having to expose `env`.

Doesn't solve the error-call issue though.

Reply:
The idea being that the quosure created through the internal `enquo()`-like call would capture and pass on the `env` for you, right? Seems interesting.

I guess that quosure env becomes the parent env if you want to insert a `.qux` pronoun or something

Reply:
yup we'd have to take `data` instead of `env` in this case. And if multiple layer masks are ever needed, we could extend to accept rlang data masks which make the layer structure explicit so we'd have all the information needed (even though we wouldn't use `eval_tidy()` but `glue()` in the end, I think that should still work).

Reply:
But the compiler would get us in trouble though, since it unwraps strings from promises and then we lose the environment. We can write this idea off.